File: src/main/java/io/playground/orderservice/infrastructure/persistence/order/entity/OrderItemEntity.java
package io.playground.orderservice.infrastructure.persistence.order.entity;

import io.playground.orderservice.domain.order.OrderItem;
import jakarta.persistence.*;
import lombok.AllArgsConstructor;
import lombok.Builder;
import lombok.Getter;
import lombok.NoArgsConstructor;

import java.math.BigDecimal;

@Entity
@Getter
@NoArgsConstructor(access = lombok.AccessLevel.PROTECTED)
@AllArgsConstructor
@Builder
@Table(
        name = "order_items",
        uniqueConstraints = {
                @UniqueConstraint(
                        name = "uk_variantId_orderId",
                        columnNames = {"variantId", "orderId"}
                )
        }
)
public class OrderItemEntity {
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    private Long id;

    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id",nullable = false)
    private OrderEntity order;

    @Column(nullable = false)
    private Long variantId;

    @Column(nullable = false)
    private String variantName;

    @Column(nullable = false)
    private Long productId;

    @Column(nullable = false)
    private String productName;

    @Column(nullable = false)
    private BigDecimal price;

    @Column(nullable = false)
    private int quantity;

    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    private OrderItem.OrderItemStatus status;

    @Column(nullable = false)
    private int canceledQuantity;

    @Column
    private String canceledReason;

    public static OrderItemEntity fromDomain(OrderItem orderItem,
                                             OrderEntity orderEntity) {
        return OrderItemEntity.builder()
                .order(orderEntity)
                .variantId(orderItem.getVariantId())
                .variantName(orderItem.getVariantName())
                .productId(orderItem.getProductId())
                .productName(orderItem.getProductName())
                .price(orderItem.getPrice())
                .quantity(orderItem.getQuantity())
                .status(orderItem.getStatus())
                .canceledQuantity(orderItem.getCanceledQuantity())
                .canceledReason(orderItem.getCanceledReason())
                .build();
    }

    public OrderItem toDomain() {
        return OrderItem.of(
                this.id,
                this.order.getId(),
                this.variantId,
                this.variantName,
                this.productId,
                this.productName,
                this.price,
                this.quantity,
                this.status,
                this.canceledQuantity,
                this.canceledReason
        );
    }
}

File: src/main/kotlin/io/playground/orderservice/infrastructure/persistence/order/entity/OrderItemEntity.kt
package io.playground.orderservice.infrastructure.persistence.order.entity

import io.playground.orderservice.domain.order.OrderItem
import jakarta.persistence.Column
import jakarta.persistence.Entity
import jakarta.persistence.EnumType
import jakarta.persistence.Enumerated
import jakarta.persistence.FetchType
import jakarta.persistence.GeneratedValue
import jakarta.persistence.GenerationType
import jakarta.persistence.Id
import jakarta.persistence.JoinColumn
import jakarta.persistence.ManyToOne
import jakarta.persistence.Table
import jakarta.persistence.UniqueConstraint
import java.math.BigDecimal

@Entity
@Table(
    name = "order_items",
    uniqueConstraints = [
        UniqueConstraint(name = "uk_variantId_orderId", columnNames = ["variantId", "orderId"]),
    ],
)
class OrderItemEntity(
    @Id
    @GeneratedValue(strategy = GenerationType.IDENTITY)
    var id: Long? = null,
    @ManyToOne(fetch = FetchType.LAZY)
    @JoinColumn(name = "order_id", nullable = false)
    var order: OrderEntity? = null,
    @Column(nullable = false)
    var variantId: Long = 0,
    @Column(nullable = false)
    var variantName: String = "",
    @Column(nullable = false)
    var productId: Long = 0,
    @Column(nullable = false)
    var productName: String = "",
    @Column(nullable = false)
    var price: BigDecimal = BigDecimal.ZERO,
    @Column(nullable = false)
    var quantity: Int = 0,
    @Enumerated(EnumType.STRING)
    @Column(nullable = false)
    var status: OrderItem.OrderItemStatus = OrderItem.OrderItemStatus.NONE,
    @Column(nullable = false)
    var canceledQuantity: Int = 0,
    @Column
    var canceledReason: String? = null,
) {
    fun toDomain(): OrderItem = OrderItem.of(
        id,
        order?.id,
        variantId,
        variantName,
        productId,
        productName,
        price,
        quantity,
        status,
        canceledQuantity,
        canceledReason,
    )

    companion object {
        @JvmStatic
        fun fromDomain(orderItem: OrderItem, orderEntity: OrderEntity): OrderItemEntity =
            OrderItemEntity(
                order = orderEntity,
                variantId = orderItem.variantId,
                variantName = orderItem.variantName,
                productId = orderItem.productId,
                productName = orderItem.productName,
                price = orderItem.price,
                quantity = orderItem.quantity,
                status = orderItem.status,
                canceledQuantity = orderItem.canceledQuantity,
                canceledReason = orderItem.canceledReason,
            )
    }
}
